File: python/04/star_two.py
import os

input_file = os.path.join(os.path.dirname(__file__), "../../input/day_04.txt")

total = {x: 1 for x in range(196)}  # too lazy to refactor
with open(input_file, "r") as file:
    for idx, x in enumerate(file.readlines()):
        x = x[:-1]
        win_nums, own_nums = x.split(":")[1].split("|")
        win_nums = set(win_nums.split(" "))
        win_nums.remove("")

        match_count = 0
        for y in own_nums.split(" "):
            if y in win_nums:
                match_count += 1
        
        for i in range(idx + 1, idx + match_count + 1):
            total[i] += total[idx]

print(sum(total.values()))

File: typescript/04/star_two.ts
import * as fs from "fs/promises";
import * as path from "path";

(async () => {
    const rawInput = await fs.readFile(path.resolve(__dirname, '../../input/day_04.txt'), "utf-8");
    const arr = rawInput.split("\n");

    const totalDict: { [key: number]: number } = {};
    for (let x = 0; x < 196; x++) {
        totalDict[x] = 1;
    }

    for (let idx = 0; idx < arr.length; idx++) {
        const [winNumsTmp, ownNums] = arr[idx].split(":")[1].split("|");
        const winNums = winNumsTmp.split(" ").filter(value => value != "")
        
        let matchCount = 0
        ownNums.split(" ").forEach(y => {
            if (winNums.includes(y))
                matchCount ++;
        })

        for (let i = idx + 1; i <= idx + matchCount; i++) {
            totalDict[i] += totalDict[idx]            
        }
    }
    const total = Object.values(totalDict).reduce((sum, num) => sum + num)
    console.log(total)
})();